File: Dec2020/5Dec/LemonadeChange.java
/*
Problem Name: 860. Lemonade Change
Problem Link: https://leetcode.com/problems/lemonade-change/
*/
class Solution {
    public boolean lemonadeChange(int[] bills) {
        int five = 0, ten = 0;
        for(int i=0;i<bills.length;i++){
            if(bills[i] == 5)
                five++;
            else if(bills[i] == 10){
                if(five == 0)
                    return false;
                five--;
                ten++;
            }
            else{
                if(ten == 0){
                    if(five < 3)
                        return false;
                    five -= 3;
                }
                else{
                    if(five == 0)
                        return false;
                    five--;
                    ten--;
                }   
            }
        }
        return true;
    }
}

File: Dec2020/5Dec/LemonadeChange.cpp
/*
Problem name: 860. Lemonade Change
Problem link: https://leetcode.com/problems/lemonade-change/
*/

class Solution {
public:
    bool lemonadeChange(vector<int>& bills) {
        bool change = true;
        int arr[2]={0};   // 0 represents 5Rs note; 1 - 10 Rs; 2 - 15Rs.
        for(int i=0;i<bills.size();++i){
            if(bills[i] == 5)
                arr[0]++;
            
            else if(bills[i] == 10)
                if(arr[0]>0){
                    arr[0]--;
                    arr[1]++;
                }
                else{
                    change = false;
                    break;
                }
            
            else
                if(arr[1]>0 && arr[0]>0){
                    arr[1]--;
                    arr[0]--;
                }
                else if(arr[0] > 2){
                    arr[0]--;
                    arr[0]--;
                    arr[0]--;
                }
                else{
                    change = false;
                    break;
                }
        }
        
        if(change)  cout<<"true";
        else    cout<<"false";
        
        return change;
    }
};
